### ISSUE TITLE
Release version of the Linux plugin crashes

### ISSUE THREAD

**Issue by jsakkine** [surge-synthesizer/surge#469]
Debug version works but the release version crashes somewhere in the function load_mid_controllers().

**Comment by jsakkine** [comment]
Noticed while working on my resource manager. The issue has predated that one.

**Comment by baconpaul** [comment]
Yeah I saw the same when I tried the release version; agree this is not your changes.

**Comment by baconpaul** [comment]
OK I know why this is happening and it is because of a very evil non standard thing in the xml parser. I actually trued to fix this when I did the mac port and it broke a lot of stuff.

load_mid_controllrs (and elsewhere) has code which looks like this

```
   TiXmlElement* mc = getSnapshotSection("midictrl");
   assert(mc);
   
   TiXmlElement* entry = mc->FirstChild("entry")->ToElement():
```

Now you may look at that and say "hold on; how can mc->FirstChild("entry")->ToElement be safe if mc->FirstChild("entry") is null"?

And you would be right.

But the people who wrote tinyxml thought about solving (or should I say "solving") that in a nasty way

```
	TiXmlElement*  ToElement()			{ return ( this && type == ELEMENT  ) ? (TiXmlElement*)  this : 0; } ///< Cast to a more defined type. Will return null not of the requested type.
```

That's right. ToElement actually checks that 'this' is never null.

Now on mac when you compile you get a bunch of warnings saying "this can never be null and it may be ignored". Because, hey, this can never be null and it may be ignored.

So when gcc -O3 runs, it knows this can never be null and branch jumps directly to side 2 of the &&

and then dumps core

because that code is wrong

but is used everywhere

So fun. Our choices are

```
TiXmlElement* entry = (mc->FirstChild("entry")?mc->FirstChild("entry")->ToElement():NULL);
```

lines like that through SurgeStorage (which I've done everywhere) or alternately do something clever with a return type other than NULL which is == NULL and replace the this check with it.

But that's your problem.

It is blind luck, by the way, that llvm on mac and cl on windows don't optimize that away. One day a compiler upgrade could just as easily make mac or windows core dump.

Ugh.

**Comment by baconpaul** [comment]
The easiest (but least pleasant) solution is to introduce a macro TO_ELEMENT which SurgeStorage, CSnapshotMenu and SurgePatch all have which is

```
#define TO_ELEMENT(expr) (expr?expr->ToElement():NULL)
```

and then change the ->ToElement calls in the code.

While we all don't love that fix, if we did it I would approve the PR. Smallest change we can make even though it leaves a lurking problem in the third party xml code.

**Comment by baconpaul** [comment]
Or even better `dynamic_cast<TiXmlElement *>(expr->FirstChild("Entry"))` which is basically what the code is doing

**Comment by baconpaul** [comment]
OK so I have a slightly different proposal than the macro now

1: in lib/xml/tinyxml.h ifdef out line 606-619 so using ->ToElement won't compile for anyone
2: Where the code breaks, replace it with a dynamic_cast
3: Test a bit, push.

wdyt?

**Comment by baconpaul** [comment]
Or alternately write a cast operator which checks null on the argument and take it out of being a member. Either one would work.

**Comment by baconpaul** [comment]
Yeah this diff pushes the crash later. (presumably to another xml point)

```
diff --git a/src/common/SurgeStorage.cpp b/src/common/SurgeStorage.cpp
index e531635..1271401 100644
--- a/src/common/SurgeStorage.cpp
+++ b/src/common/SurgeStorage.cpp
@@ -213,7 +213,7 @@ SurgeStorage::SurgeStorage()
 #endif
    }
 
-   TiXmlElement* e = snapshotloader.FirstChild("autometa")->ToElement();
+   TiXmlElement* e = dynamic_cast<TiXmlElement*>(snapshotloader.FirstChild("autometa"));
    if (e)
    {
       defaultname = e->Attribute("name");
@@ -891,7 +891,7 @@ void SurgeStorage::load_midi_controllers()
    TiXmlElement* mc = getSnapshotSection("midictrl");
    assert(mc);
 
-   TiXmlElement* entry = mc->FirstChild("entry")->ToElement();
+   TiXmlElement* entry = dynamic_cast<TiXmlElement*>(mc->FirstChild("entry"));
    while (entry)
    {
       int id, ctrl;
@@ -902,13 +902,13 @@ void SurgeStorage::load_midi_controllers()
          if (id >= n_global_params)
             getPatch().param_ptr[id + n_scene_params]->midictrl = ctrl;
       }
-      entry = entry->NextSibling("entry")->ToElement();
+      entry = dynamic_cast<TiXmlElement*>(entry->NextSibling("entry"));
    }
 
    TiXmlElement* cc = getSnapshotSection("customctrl");
    assert(cc);
 
-   entry = cc->FirstChild("entry")->ToElement();
+   entry = dynamic_cast<TiXmlElement*>(cc->FirstChild("entry"));
    while (entry)
    {
       int id, ctrl;
@@ -917,7 +917,7 @@ void SurgeStorage::load_midi_controllers()
       {
          controllers[id] = ctrl;
       }
-      entry = entry->NextSibling("entry")->ToElement();
+      entry = dynamic_cast<TiXmlElement*>(entry->NextSibling("entry"));
    }
 }
```

**Comment by baconpaul** [comment]
And yup my next crash is in CSnapshotMenu.

I'm 3/4 the way done. Want me to make a branch and PR this in? I can also get the log2 thing while I'm at it.

**Comment by jsakkine** [comment]
Ugh, great analysis and how crappy code...

**Comment by jsakkine** [comment]
I mean crappy code in TinyXml. Yes, you should explicitly check whether FirstChild() is NULL. This is really not a solid reason to bring RTTI.

**Comment by jsakkine** [comment]
Six if-statements. That is not too bad.
